### PR TITLE
[Refactor] Cleanout the logic associated to replacing the ddp sampler 

### LINKED ISSUE #9655: Sampler is being replaced even with trainer.replace_sampler_ddp=False
## 🐛 Bug

In the following line:

https://github.com/PyTorchLightning/pytorch-lightning/blob/4bebf82f271fb577ed41831403551f0bff97651b/pytorch_lightning/trainer/data_loading.py#L152

PTL tries to replace the sampler object even if `trainer.replace_sampler_ddp` is `False`:

https://github.com/PyTorchLightning/pytorch-lightning/blob/4bebf82f271fb577ed41831403551f0bff97651b/pytorch_lightning/trainer/data_loading.py#L135

### To Reproduce

Should be easy to reproduce with any custom Sampler class that has different init args from the PyTorch default.

### Expected behavior

Function in line 152 should not be called and moved just below 147. This is how it was in version v1.4.8:

https://github.com/PyTorchLightning/pytorch-lightning/blob/1.4.8/pytorch_lightning/trainer/data_loading.py#L143

### Environment

- PyTorch Lightning Version (main):
- PyTorch Version (e.g., 1.10)
- Python version: 3.8.10
- OS (e.g., Linux): Linux
- CUDA/cuDNN version: N/A
- GPU models and configuration: N/A
- How you installed PyTorch (`conda`, `pip`, source): source
- If compiling from source, the output of `torch.__config__.show()`:
- Any other relevant information:

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Fixes #9655

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/data_loading.py [hunk 1]
```diff
@@ -119,172 +112,59 @@ def auto_add_worker_init_fn(self, dataloader: DataLoader) -> None:
...
+            and self.accelerator_connector.is_distributed
+            and not isinstance(dataloader.sampler, DistributedSampler)
+            and not has_iterable_dataset(dataloader)
+        )
+
+    def prepare_dataloader(self, dataloader: Any, shuffle: bool, mode: Optional[RunningStage] = None) -> Any:
+        """This function handles to following functionalities:
+
+        - Injecting a `DistributedDataSampler` within the `DataLoader if accelertor is is_distributed
+        - Wrapping the datasets and samplers into Fault Tolerant Components
```
Comment (carmocca) [inline pytorch_lightning/trainer/data_loading.py:127]:
```suggestion
        - Wrapping the datasets and samplers into fault-tolerant components
```

File: pytorch_lightning/trainer/data_loading.py [hunk 2]
```diff
@@ -119,172 +112,59 @@ def auto_add_worker_init_fn(self, dataloader: DataLoader) -> None:
...
+                dataloader, shuffle, mode=mode, overfit_batches=self.overfit_batches, **self.distributed_sampler_kwargs
+            )
         else:
-            # use current sampler
             sampler = dataloader.sampler
 
-        dataloader = self.replace_sampler(dataloader, sampler, mode=mode)
+        # the DataLoader should be re-created only if we need to inject
+        # the fault tolerant components or the distributed sampler.
+        if _fault_tolerant_training() or isinstance(sampler, DistributedSampler):
```
Comment (carmocca) [inline pytorch_lightning/trainer/data_loading.py:156]: Isn't the condition: `or isinstance(sampler, DistributedSampler)` wrong when the above if resolved to the `else` clause and the user already set a `DistributedSampler`?

We wouldn't need to call this in that case, right?

File: pytorch_lightning/utilities/dataloader_preparation.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
+# Copyright The PyTorch Lightning team.
```
Comment (carmocca) [inline pytorch_lightning/utilities/dataloader_preparation.py:1]: Why does this code need to be moved into an utility?

IMO it's better to have these things together and we don't want to bloat the utilities.

Additionally, moving it messes up the git history so unless there's is a good reason, I wouldn't do it

File: pytorch_lightning/trainer/data_loading.py [hunk 4]
```diff
@@ -119,172 +112,59 @@ def auto_add_worker_init_fn(self, dataloader: DataLoader) -> None:
...
-                dl_kwargs["dataset"] = CaptureMapDataset(dataset=dl_kwargs["dataset"])
-            else:
-                raise MisconfigurationException(
-                    "This shouldn't happen, please open an issue on Lightning Github repository."
-                )
-
-        return dl_kwargs
-
-    @staticmethod
-    def replace_sampler(dataloader: DataLoader, sampler, mode: Optional[RunningStage] = None) -> DataLoader:
```
Comment (carmocca) [inline pytorch_lightning/trainer/data_loading.py:272]: Somebody might be using this, just pointing it out.

File: pytorch_lightning/utilities/dataloader_preparation.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/utilities/dataloader_preparation.py:1]: data loading is quite complex and I think grouping the utilities dataloader would help to understand the code better.

Comment (carmocca) [reply pytorch_lightning/utilities/dataloader_preparation.py:1]: I'd argue that grouping the utilities in the same file `data_loading.py` is more readable since you avoid the file jump. It also makes patching and subclassing easier.

File: tests/trainer/test_supporters.py [hunk 5]
```diff
@@ -325,22 +332,46 @@ def __len__(self):
...
+        dataloader = trainer.prepare_dataloader(dataloader, shuffle=True)
+        _count = 0
 
     def _assert_distributed_sampler(v):
         nonlocal _count
         _count += 1
-        assert isinstance(v, DistributedSampler)
+        if use_fault_tolerant:
+            assert isinstance(v, FastForwardSampler)
+            v = v._sampler
```
Comment (awaelchli) [inline tests/trainer/test_supporters.py:361]: does this do anything?

**Review by awaelchli — APPROVED** [review]

File: tests/trainer/test_supporters.py (hunk 5 above)
Comment (tchaton) [reply tests/trainer/test_supporters.py:364]: Yes, it extracts the wrapped sampler to make sure it is a distributed sampler. Added a check to make sure we pass within the part of the code.

File: pytorch_lightning/utilities/dataloader_preparation.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/utilities/dataloader_preparation.py:1]: Fine ! Done !

File: pytorch_lightning/trainer/data_loading.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/trainer/data_loading.py]: Done !

File: pytorch_lightning/trainer/data_loading.py [hunk 6]
```diff
@@ -67,224 +67,6 @@ class TrainerDataLoadingMixin(ABC):
     dev_debugger: InternalDebugger
     call_hook: Callable
 
-    def _worker_check(self, dataloader: DataLoader, name: str) -> None:
```
Comment (carmocca) [inline pytorch_lightning/trainer/data_loading.py:70]: Can you put the code where it was? This diff is very hard to review.

If moving the position of the code was intended, then it should be in 2 PRs. 1 modifying the code and another one moving it
